Fix Apply Changes is lost when query is edited

gabrieldutra opened the pull request:
## What type of PR is this? (check all applicable)
<!-- Please leave only what's applicable -->

- [x] Bug Fix

## Description
When editing a query the "Apply Changes" state was lost:
![state-lost](https://user-images.githubusercontent.com/3356951/61918542-1c4ef500-af28-11e9-83f6-612646a576aa.gif)



## Related Tickets & Documents
--

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
![apply-all-edit-query](https://user-images.githubusercontent.com/3356951/61919675-0c85df80-af2d-11e9-848a-10919b8b392f.gif)

ranbena commented on client/app/components/ParameterValueInput.jsx:
```diff
@@ -39,16 +39,19 @@ export class ParameterValueInput extends React.Component {
     // if value prop updated, reset dirty state
-    if (prevProps.value !== value) {
-      this.setState({ value, isDirty: false });
+    if (prevProps.value !== value || prevProps.parameter !== parameter) {
```
I think that now there's no actual need for any condition, right?

gabrieldutra replied:
It is also called when state is update, removing it would cause an infinite loop :slightly_smiling_face:

gabrieldutra commented on client/app/components/ParameterValueInput.jsx:
```diff
@@ -39,16 +39,19 @@ export class ParameterValueInput extends React.Component {
-      this.setState({ value, isDirty: false });
+    if (prevProps.value !== value || prevProps.parameter !== parameter) {
+      this.setState({
+        value: parameter.pendingValue || value,
```
Forgot those values existed :dizzy_face:

gabrieldutra commented on client/app/components/ParameterValueInput.jsx:
```diff
@@ -39,16 +39,19 @@ export class ParameterValueInput extends React.Component {
     this.state = {
-      value: props.value,
-      isDirty: false,
+      value: props.parameter.pendingValue || props.value,
```
```suggestion
      value: props.parameter.hasPendingValue ? props.parameter.pendingValue : props.value,
```

gabrieldutra commented:
> I've noticed that this happens to Query Based Dropdown Lists after typing in the query editor:

This also happens in master, the issue is that the `parameters` DOM is reconstructed every time the objects are changed (from my tests the constructor of `ParameterValueInput` was called, not `componendDidUpdate`). I changed the approach to avoid this in f81650dd728d739784ec1a5c25d26b3af7051ed1. 

I also removed the initialization of the pendingValue (although I think it should be there), because parameters could have this prop saved before when saving a query and with the initialization now they would be rendered with the state as well.

Will run some tests to make sure nothing breaks with it and comment back here.

gabrieldutra commented on client/app/services/query.js:
```diff
@@ -429,7 +429,8 @@ class Parameters {
     const parameters = this.query.options.parameters;
-    this.query.options.parameters = parameters.filter(parameterExists).map(p => new Parameter(p, this.query.id));
+    this.query.options.parameters = parameters.filter(parameterExists)
+      .map(p => (p instanceof Parameter ? p : new Parameter(p, this.query.id)));
```
This was introduced a long ago in #1649, the change seems not be affecting the fix :stuck_out_tongue_closed_eyes:
